experimental/oak_baremetal_channel/SPEC.MD

```diff
@@ -0,0 +1,118 @@
+# Oak Baremetal Communication Protocol Specification
+
+This document serves as a specification of the protocol used for communication
+between the Untrusted Loader and Trusted Runtime. The Trusted Runtime should be
+able to communicate with any implementation of the Untrusted Loader that
+implements this specification.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).
+
+## Method
+
+Methods are exposed by the Trusted Runtime. They MAY be remotely invoked by the
+Unrusted Loader. Each method has clearly defined parameters and return types.
+The Oak IDL is used to define the methods exposed by the Trusted Runtime.
+
+## Invocation
+
+An invocations of a Method. Each invocation consists out of a Request Message
+followed by a Response Message.
+
+### Issuing an invocation
+
+The Untrusted Loader MAY issue an invocation by sending a Request Message.
+
+The Trusted Runtime MUST NOT issue invocations. It MUST ignore any Response
+Messages it receives.
+
+### Handling an invocation
+
+The Trusted Runtime MUST respond to each Request Message with a single Response
+Message. It SHOULD respond to Request Messages in first-in-first-out order, but
+MAY prioritize some requests. The Trusted Runtime MAY issue Response Messages
+before the Request Message has been received in its entirety. Sending a Response
+Messages concludes the invocation.
+
+The Untrusted Loader MUST NOT send Response Messages. It MUST ignore any Request
+Messages it receives.
+
+## Messages
+
+A message forms half of an Invocation.
+
+### Request Message Definition
+
+A byte encoded request message MUST consist of the following seqeuential fields:
+
+- length, u32, little endian: length of the entire message, including this field
```

Comment:
There are two main reasons for the length. One is that is serves as a simple consistency check to ensure all the frames were received. The second is that it would allow us to pre-allocate a buffer for collecting all the frames at the start rather than to keep reallocating as we receive more frames.

Another possible approach is to always specify the entire message length in the length field of the frame, seeing that all frames except the last frame of the messages has a fixed length anyway.

Reply:
I think we should have the invocation_id in the C API if we allow the runtime to have interleaved requests; anyways this is fine, it's also fine to have the C API repeat the length, not a big deal (though it's redundant since it will also have to be passed as an arg to the C API which will probably have a signature of `invoke(buf, len)` or similar).

Reply:
Fair. :) I'd be inclined to keep it then, I don't see a very good option for avoiding having it here. Only ones are remove it (bad for reasons above), move it to the framing layers (adds some coupling, but maybe ok?), or add another layer on top (don't love the complexity of that).